src/Java.Interop.Tools.Cecil/Java.Interop.Tools.Cecil/NullableReferenceTypesRocks.cs

```diff
@@ -0,0 +1,121 @@
+using System;
+using System.Linq;
+using Mono.Cecil;
+using Mono.Collections.Generic;
+
+namespace Java.Interop.Tools.Cecil
+{
+	// Partial support for determining NRT status of method and field types.
+	// Reference: https://github.com/dotnet/roslyn/blob/main/docs/features/nullable-metadata.md
+	// The basics are supported, but advanced annotations like array elements,
+	// type parameters, and tuples are not supported.  Our use case doesn't really need them.
```

Comment:
At least for commit message purposes, an elaboration on "our use case doesn't really need them" would be good.  Is it not possible for Java code to express the difference between `String?[]?` vs. `String[]?` (possibly nullable array containing possibly null items vs. possibly null array which won't contain any null items)?  That's my assumption.

At the same time, I now wonder if we're binding arrays & collections *wrong*; should *all* arrays permit null items?

As a "spot-check", we're binding [`DatabaseUtils.longForQuery(SQLiteDatabase, String, String[])`](https://developer.android.com/reference/android/database/DatabaseUtils#longForQuery(android.database.sqlite.SQLiteDatabase,%20java.lang.String,%20java.lang.String[])) as:

```csharp
public static unsafe long LongForQuery (Android.Database.Sqlite.SQLiteDatabase? db, string? query, string[]? selectionArgs);
```

And in this case, that looks correct; null items aren't allowed:

  * https://github.com/aosp-mirror/platform_frameworks_base/blob/ec84453381ef4a772e37d3bbf8c9d4da560ab19c/core/java/android/database/DatabaseUtils.java#L1002-L1009
  * https://github.com/aosp-mirror/platform_frameworks_base/blob/ec84453381ef4a772e37d3bbf8c9d4da560ab19c/core/java/android/database/DatabaseUtils.java#L1016
  * https://github.com/aosp-mirror/platform_frameworks_base/blob/ec84453381ef4a772e37d3bbf8c9d4da560ab19c/core/java/android/database/sqlite/SQLiteProgram.java#L201-L207
  * https://github.com/aosp-mirror/platform_frameworks_base/blob/ec84453381ef4a772e37d3bbf8c9d4da560ab19c/core/java/android/database/sqlite/SQLiteProgram.java#L166-L171

Compare to [`Arrays.hashCode(Object[])`](https://developer.android.com/reference/java/util/Arrays#hashCode(java.lang.Object[])), which we similarly bind as:

```csharp
public static unsafe int HashCode (Java.Lang.Object[]? a);
```

i.e. array can be null, but elements can't be.  Compare to the source code, which does allow for null elements:

  * https://android.googlesource.com/platform/libcore/+/fe39951/luni/src/main/java/java/util/Arrays.java#1249

implying that we should have bound this as `Object?[]?`, not `Object[]?`.

I'm not sure there's a "good" answer here, particularly since the few Java annotations I've seen around nullability work at an "outermost type" level.

*However*, Kotlin appears to support it:

```kotlin
class E {
    fun m1(v : Array<String?>?)
    {
    }

    fun m2(v : Array<String>?)
    {
    }
}
```

…but that said, our `//@not-null` attribute (01d06845499794601c2beedca50e1c5b2fa4caa3) only works on the "outermost type" level as well, meaning we can't differentiate `Array<String?>` from `Array<String?>?` *anyway*…

Reply:
Technically some of the annotations we look for do support `string?[]?`, like `Lorg/jetbrains/annotations/NotNull;`:

```java
package org.jetbrains.annotations;

@Target({ElementType.METHOD, ElementType.FIELD, ElementType.PARAMETER, ElementType.LOCAL_VARIABLE, ElementType.TYPE_USE})
public @interface NotNull { ... }
```

([source](https://github.com/JetBrains/java-annotations/blob/master/common/src/main/java/org/jetbrains/annotations/NotNull.java))

(My understanding is it's the `ElementType.TYPE_USE` that allows specifying `string?[]`.)

However the Android ones we primarily use do not support that:
```java
package androidx.annotation;

@Target({METHOD, PARAMETER, FIELD, LOCAL_VARIABLE, ANNOTATION_TYPE, PACKAGE})
public @interface NonNull { ... }
```

([source](https://android.googlesource.com/platform/frameworks/support/+/refs/tags/android-9.0.0_r40/annotations/src/main/java/androidx/annotation/NonNull.java))

As you pointed out, ***our*** tools do not support specifying the array element nullability.

While we probably wouldn't gain much by adding support to our tools since the primary Android annotations don't support it, we probably should update `generator` to always emit nullable array elements, since the default should be nullable unless specified `@NotNull`.

So we should always emit `string?[]` or `string?[]?`.